components/gitpod-protocol/src/util/tracing.ts

```diff
@@ -76,15 +81,85 @@ export namespace TraceContext {
             return;
         }
 
-        ctx.span.addTags({
+        addNestedTags(ctx, {
             rpc: {
                 system: "jsonrpc",
                 method,
                 jsonrpc: {
-                    parameters: args.slice(),
+                    // parameters are added below, because we don't want them be nested, so that we have the possiblity to redact them.
                 },
             },
         });
+
+        for (let i = 0; i < args.length; i++) {
+            ctx.span.setTag(`rpc.jsonrpc.parameters.${i}`, args[i]);
+        }
+    }
+
+    /**
+     * Offers a way to opt-out of tracing for a particular JsonRPC parameter. Required if it contains PII, for instance.
+     * @param ctx
+     * @param indexes
+     * @returns
+     */
+    export function redactJsonRPCParameters(ctx: TraceContext, indexes: number[]) {
+        if (!ctx.span) {
+            return;
+        }
+
+        for (const i of indexes) {
+            ctx.span.setTag(`rpc.jsonrpc.parameters.${i}`, "REDACTED");
+        }
+    }
```

Comment:
I can share some of my thoughts, but hope @mads-hartmann has some thoughts and experience to share as well! :smile: 

Regarding PII in general, especially the question "what if this harmless field contains PII" (example: `but what if layoutData accidentally contains PII?`)
 - PII is about the fields that are semantically known to contain PII, like `name`, `address`, `githubUserId`, which can be used (in combination with publicly available sources, like GitHub) to identify you in person
 - it's _not_ about user (mis-)using other fields to put their name, address, etc. Imagine a repo named `<myfullname>_<myaddress>`: we cannot be held accountable to censor this; that's impossible
 - that being said, there might be cases where we'd like to err on the side of caution. But that again is independent from "opt-in vs opt-out"; as our goal is to trace as much as possible we have to decide in either case. `layoutData` might be such a case; I personally have no clue what's in there. :shrug: 

Regarding the general choice "opt-in vs opt-out":
 - I personally have no hard feelings here. I started this PR with opt-in because of [this comment](https://github.com/gitpod-io/gitpod/pull/6995#pullrequestreview-820266817) from @mads-hartmann how has some experience with tracing. Having a single line at the beginning of each method is also fine with me. Would safe us from the strange "indentify by index" problem (cmp. `redactJsonRPCParameters`)
 - to re-iterate:
   - we have to be careful with PII either way, every time we change the API (nested shapes)
   - it's desirable to trace as much as possible (as long as it's not "huge data"). This is different to before, where we'd be cautious with what we trace because how badly our self-managed jaeger handled the (small) amount of data. So comparing to "how it was before" is not a good guideline: traces where near-unusable, and not present for most API calls in the first place. :slightly_smiling_face: 

> UserEnvVarValue in setEnvVar should probably always be censored

> GuessGitTokenScopesParams contains a GitToken

Both true :100:, nice findings. :pray: Will fix.

> Is the content part of UpdateUserStorageResourceOptions always safe to trace in updateUserStorageResource?

Content is already excluded I think. Will verify.

Reply:
I can extend a bit about the motivations behind adding all parameters automatically, if you don't mind a little detour.

There isn't a single agreed-upon definition of observability. Back in [august 2019](https://blog.mads-hartmann.com/sre/2019/08/04/journey-into-observability-reading-material.html) I had the following take (sorry for quoting myself 😂) based on what I had read and my experience at the time.

> Observability is all about being able to ask questions of your system and get answers based on the existing telemetry it produces; if you have to re-configure or modify the service to get answers to your questions you haven’t achieved observability yet

I think it's still a good guiding definition to consider when you instrument your services. You want to strive towards having all the information you need available to you when you're debugging, and sometimes you can't know what that information is ahead of time. With metrics this is impossible as each label you add will introduce more time series which explodes based on the number of labels and their cardinality. With traces that's not a concern, so you're free to add as many attributes as you like. So I'd always default to add things automatically assuming that you can namespace it correctly.

So in that sense, having to specify exactly which parameter to add to the traces means you have to think about what information you might need during debugging ahead of time, and you will inevitably overlook something important.

The PII concern is important though. Honeycomb has a retention of 60 so the data is gone after that, but it would still have reached their servers of course. 

It's a shame that JSON RPC arguments are positional, otherwise we might have been able to add some processing in the OpenTelemetry Collector to redact any parameter name that matched "email" or "address" etc.

I'd lean towards the opt-out approach as the vast majority of your parameters (I'm assuming) isn't PII. And as you already have to be careful about PII I don't think it's too much additional load to think about redacting it from telemetry.

But this is up to you. Having a single line in each method also allows you to name the parameters as @geropl highlighted, so that has benefits too. Either way this is a huge improvement and I'm happy to see it.

Reply:
After having read through the rest of the comments here I agree with @jankeromnes that redacting based on the position of the argument is fragile and I change my recommendation to be that you add a line where you opt-in to adding the arguments, which also gives you the opportunity to give them names, as you suggested @geropl ☺️

For the future, perhaps we can find a way to use named arguments with JSON RPC so we can get the best of both worlds.

Reply:
Thx @mads-hartmann ! :+1: make sense.


> For the future, perhaps we can find a way to use named arguments with JSON RPC so we can get the best of both worlds.

Hopefully we can replace it with something else; maybe gRPC. :smiley: